### PR TITLE
[SYCL-MLIR][LoopInternalization] Add limitations on candidates

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Only transforms kernel body function with `nd_item` argument.

### CODE REVIEW

File: polygeist/test/polygeist-opt/sycl/loop_internalization.mlir [hunk 1]
```diff
@@ -1,84 +1,124 @@
...
 // CHECK-NEXT:  }
-func.func @affine_1d() {
+gpu.module @device_func {
+func.func private @affine_1d(%arg0: memref<?x!sycl_accessor_1_f32_r_gb>, %arg1: memref<?x!sycl_nd_item_1_>) {
   affine.for %i = 0 to 256 {
     "test.foo"(%i) : (index) -> ()
   }
   return
 }
+gpu.func @kernel(%arg0: memref<?x!sycl_accessor_1_f32_r_gb>, %arg1: memref<?x!sycl_nd_item_1_>) kernel {
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/sycl/loop_internalization.mlir:35]: > How realistic is it that the SYCL kernel function has this signature at any stage in our actual pipeline?

The signature of the kernel is not realistic, but the signature of the callee is realistic, and that's what the pass look at. 
The pass doesn't rely on the signature of the kernel. Wanted to simplify the lit tests to not construct nd_item in the kernel before calling the callee.

> Would we potentially miss candidates if we rely on this and the kernel does not call another function, but contains the loop directly in its body?

With the current pipeline, inlining is not run until after detect reduction, so loops are unlikely (or impossible) to be directly in its body.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLTypes.h [hunk 2]
```diff
@@ -49,11 +49,8 @@ llvm::SmallVector<mlir::TypeID> getDerivedTypes(mlir::TypeID typeID);
 namespace mlir {
 namespace sycl {
 
-/// Return true if type is 'memref<?x!sycl.accessor>'.
-bool isAccessorPtrType(Type type);
-
-/// Return true if type is 'memref<?x!sycl.id>'.
-bool isIDPtrType(Type type);
+/// Return true if type is 'memref<?xT>'.
+template <typename T> bool isPtrType(Type type);
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLTypes.h]: done.

File: polygeist/test/polygeist-opt/sycl/loop_internalization.mlir (hunk 1 above)
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/test/polygeist-opt/sycl/loop_internalization.mlir:35]: > With the current pipeline, inlining is not run until after detect reduction, so loops are unlikely (or impossible) to be directly in its body.

I'm not sure I can follow you here. 

What would happen with following SYCL kernel?
```
cgh.parallel_for<class LoopKernel>(someNDRange, [&](sycl::nd_item i){
  for(...){
    for(...){
      ...
    }
  }
});
```

Even without inlining, wouldn't the loops be directly in the body of the kernel and we would miss this candidate?

**Review by sommerlukas — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 3]
```diff
@@ -48,6 +49,18 @@ namespace {
 // Utilities functions
 //===----------------------------------------------------------------------===//
 
+// Only visit kernel body function with nd_item argument.
+bool isCandidate(FunctionOpInterface func) {
+  if (!polygeist::isPotentialKernelBodyFunc(func))
+    return false;
+
+  if (func.getNumArguments() == 0 ||
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:58]: done.

**Pull request merged**